crates/typst/src/util/str.rs

```diff
@@ -0,0 +1,63 @@
+use ecow::EcoString;
+use lasso::{Spur, ThreadedRodeo};
+use once_cell::sync::Lazy;
+use typst_macros::cast;
+
+/// The global string interner.
+static INTERNER: Lazy<ThreadedRodeo> = Lazy::new(ThreadedRodeo::new);
+
+/// An interned string.
+///
+/// The API is purposefully kept small. This is because it might be relatively
+/// slow to look up a string in the interner, so we want to avoid doing it
+/// unnecessarily. For this reason, the user should use the [`PicoStr::resolve`]
+/// method to get the underlying string, such that the lookup is done only once.
+#[derive(Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct PicoStr(Spur);
+
+impl PicoStr {
+    /// Creates a new interned string.
+    pub fn new<S: AsRef<str>>(s: S) -> Self {
```

Comment:
```suggestion
    pub fn new(s: impl AsRef<str>) -> Self {
```

Reply:
You younglings and your `impl Trait`, back in my days we didn't have 'dem fancy features. /s

Fixed ;)

Reply:
Back in my days, we didn't either. But I go with the times. :)